File: api/utils/files.js
const fs = require('fs');
const path = require('path');
const { FILE_RETENTION_MS, DISK_SPACE_MIN_GB } = require('../config/constants');

const TEMP_DIR = '/var/tmp/yoink';
const TEMP_DIRS = {
  download: path.join(TEMP_DIR, 'downloads'),
  convert: path.join(TEMP_DIR, 'convert'),
  compress: path.join(TEMP_DIR, 'compress'),
  playlist: path.join(TEMP_DIR, 'playlists'),
  gallery: path.join(TEMP_DIR, 'galleries'),
  upload: path.join(TEMP_DIR, 'uploads'),
  bot: path.join(TEMP_DIR, 'bot'),
  transcribe: path.join(TEMP_DIR, 'transcribe')
};

function clearTempDir() {
  Object.values(TEMP_DIRS).forEach(dir => {
    try {
      if (fs.existsSync(dir)) {
        const items = fs.readdirSync(dir);
        items.forEach(item => {
          const itemPath = path.join(dir, item);
          try {
            const stat = fs.statSync(itemPath);
            if (stat.isDirectory()) fs.rmSync(itemPath, { recursive: true });
            else fs.unlinkSync(itemPath);
          } catch {}
        });
      } else {
        fs.mkdirSync(dir, { recursive: true });
      }
    } catch {
      fs.mkdirSync(dir, { recursive: true });
    }
  });
  console.log('✓ Cleared temp directories');
}

function cleanupTempFiles() {
  try {
    const now = Date.now();
    Object.values(TEMP_DIRS).forEach(dir => {
      if (!fs.existsSync(dir)) return;
      const items = fs.readdirSync(dir);
      items.forEach(item => {
        const itemPath = path.join(dir, item);
        try {
          const stat = fs.statSync(itemPath);
          if (now - stat.mtimeMs > FILE_RETENTION_MS) {
            if (stat.isDirectory()) {
              fs.rmSync(itemPath, { recursive: true });
            } else {
              fs.unlinkSync(itemPath);
            }
            console.log(`Cleaned up old temp: ${item}`);
          }
        } catch { }
      });
    });
    try {
      const stats = fs.statfsSync(TEMP_DIR);
      const availableGB = (stats.bavail * stats.bsize) / (1024 * 1024 * 1024);
      const totalGB = (stats.blocks * stats.bsize) / (1024 * 1024 * 1024);
      const usedGB = totalGB - availableGB;
      console.log(`[DiskSpace] ${availableGB.toFixed(1)}GB free / ${totalGB.toFixed(1)}GB total (${usedGB.toFixed(1)}GB used)`);
      if (availableGB < DISK_SPACE_MIN_GB) {
        console.warn(`[DiskSpace] WARNING: Only ${availableGB.toFixed(1)}GB free, below ${DISK_SPACE_MIN_GB}GB threshold!`);
      }
    } catch {}
  } catch (err) {
    console.error('Cleanup error:', err);
  }
}

function cleanupJobFiles(jobId) {
  try {
    let cleaned = 0;
    Object.values(TEMP_DIRS).forEach(dir => {
      if (!fs.existsSync(dir)) return;
      const items = fs.readdirSync(dir);
      items.forEach(item => {
        if (item.includes(jobId)) {
          const itemPath = path.join(dir, item);
          try {
            const stat = fs.statSync(itemPath);
            if (stat.isDirectory()) {
              fs.rmSync(itemPath, { recursive: true });
            } else {
              fs.unlinkSync(itemPath);
            }
            console.log(`[Cleanup] Removed: ${item}`);
            cleaned++;
          } catch (innerErr) {
            console.debug(`[Cleanup] Failed to remove ${item} for job ${jobId}: ${innerErr.message}`);
          }
        }
      });
    });
    if (cleaned === 0) {
      console.debug(`[Cleanup] No files found for job ${jobId.slice(0, 12)}`);
    }
  } catch (outerErr) {
    console.debug(`[Cleanup] Failed to cleanup job ${jobId}: ${outerErr.message}`);
  }
}

function sanitizeFilename(filename) {
  return filename
    .replace(/[<>:"/\\|?*\x00-\x1f]/g, '_')
    .replace(/\s+/g, ' ')
    .trim()
    .slice(0, 200);
}

function startCleanupInterval() {
  setInterval(cleanupTempFiles, 5 * 60 * 1000);
}

module.exports = {
  TEMP_DIR,
  TEMP_DIRS,
  clearTempDir,
  cleanupTempFiles,
  cleanupJobFiles,
  sanitizeFilename,
  startCleanupInterval
};

File: internal/util/files.go
package util

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/coah80/yoink/internal/config"
)

var unsafeFilenameRe = regexp.MustCompile(`[<>:"/\\|?*\x00-\x1f]`)
var multiSpaceRe = regexp.MustCompile(`\s+`)

func ClearTempDir() {
	for _, dir := range config.TempDirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			os.MkdirAll(dir, 0755)
			continue
		}
		for _, e := range entries {
			p := filepath.Join(dir, e.Name())
			os.RemoveAll(p)
		}
	}
	fmt.Println("✓ Cleared temp directories")
}

func CleanupTempFiles() {
	now := time.Now()
	for _, dir := range config.TempDirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			p := filepath.Join(dir, e.Name())
			info, err := e.Info()
			if err != nil {
				continue
			}
			if now.Sub(info.ModTime()) > config.FileRetention {
				os.RemoveAll(p)
				log.Printf("Cleaned up old temp: %s", e.Name())
			}
		}
	}

	if ds, err := GetDiskSpace(config.TempDir); err == nil {
		log.Printf("[DiskSpace] %.1fGB free / %.1fGB total (%.1fGB used)", ds.AvailGB, ds.TotalGB, ds.UsedGB)
		if ds.AvailGB < float64(config.DiskSpaceMinGB) {
			log.Printf("[DiskSpace] WARNING: Only %.1fGB free, below %dGB threshold!", ds.AvailGB, config.DiskSpaceMinGB)
		}
	}
}

func CleanupJobFiles(jobID string) {
	cleaned := 0
	for _, dir := range config.TempDirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			if strings.Contains(e.Name(), jobID) {
				p := filepath.Join(dir, e.Name())
				if err := os.RemoveAll(p); err == nil {
					log.Printf("[Cleanup] Removed: %s", e.Name())
					cleaned++
				}
			}
		}
	}
	if cleaned == 0 {
		short := jobID
		if len(short) > 12 {
			short = short[:12]
		}
		log.Printf("[Cleanup] No files found for job %s", short)
	}
}

func SanitizeFilename(filename string) string {
	s := unsafeFilenameRe.ReplaceAllString(filename, "_")
	s = multiSpaceRe.ReplaceAllString(s, " ")
	s = strings.TrimSpace(s)
	if len(s) > 200 {
		s = s[:200]
	}
	return s
}

func StartCleanupInterval() {
	ticker := time.NewTicker(5 * time.Minute)
	go func() {
		for range ticker.C {
			CleanupTempFiles()
		}
	}()
}
